posts/python_sql_engine.md

```diff
@@ -0,0 +1,187 @@
+
+Finally, the components that have been built to support execution can be used as a **foundation** for a faster engine. I'm inspired by what [Apache Calcite](https://github.com/apache/calcite) has done for the JVM world. Even though Python is commonly used for data, there hasn't been a Calcite for Python. So you could say that SQLGlot aims to be that framework. For example, it wouldn't take much work to replace the Python execution engine with numpy / pandas / arrow to become a respectably performing query engine. The implementation would be able to leverage the parser, optimizer, and logical planner, only needing to implement physical execution. There is a ton of work in the Python ecosystem around high performance vectorized computation which I think could benefit from a pure Python based [AST](https://en.wikipedia.org/wiki/Abstract_syntax_tree) / [plan](https://en.wikipedia.org/wiki/Query_plan). Parsing and planning doesn't have to be fast when the bottleneck of running queries is processing terabytes of data. So having a Python based ecosystem around SQL is beneficial given the ease of development in Python despite not having bare metal performance.
+
+Parts of SQLGlot's toolkit are being used today by [Ibis](https://github.com/ibis-project/ibis), [Quokka](https://github.com/marsupialtail/quokka), and [Splink](https://github.com/moj-analytical-services/splink).
+
+## How?
+
+There are many steps involved with actually running a simple query like:
+
+```sql
+SELECT
+  bar.a,
+  b + 'y' AS b
+FROM bar
+JOIN baz
+  ON bar.a = baz.a
+WHERE bar.a > 1
+```
+
+In this post, I'll walk through all the steps SQLGlot takes to run this query over Python objects.
+
+## Tokenizing
+
+The first step is to convert the sql string into a list of tokens. SQLGlot's tokenizer is quite simple and can be found [here](https://github.com/tobymao/sqlglot/blob/main/sqlglot/tokens.py#L763). In a while loop, it checks each character and either appends the character to the current token or makes a new token.
+
+Running the SQLGlot tokenizer shows the output.
+
+![Tokenizer Output](python_sql_engine_images/tokenizer.png)
+
+Each keyword has been converted to a SQLGlot Token object. Each token has some metadata associated with it like line / column information for error messages. Comments are also a part of the token so that comments can be preserved.
```

Comment:
```suggestion
Each keyword has been converted to a SQLGlot Token object. Each token has some metadata associated with it like line / column information for error messages. Comments are also a part of the token so that they can be preserved when transpiling SQL.
```

Reply:
The sentence about comments doesn't feel necessary.